src/launch/mod.rs

```diff
@@ -192,3 +194,171 @@ impl codicon::Encoder<()> for Measurement {
+    }
+}
+
+/// Describes a policy that the AMD Secure Processor will
+/// enforce.
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, Deserialize, Serialize)]
+pub struct SnpPolicy {
+    /// The various policy optons are encoded as bit flags.
+    pub flags: SnpPolicyFlags,
+
+    /// The desired minimum platform firmware version.
+    pub minfw: Version,
+}
+
+impl SnpPolicy {
+    /// Convert a Policy to it's u64 counterpart.
+    pub fn as_u64(&self) -> u64 {
+        let mut val: u64 = 0;
+
+        let minor_version = u64::from(self.minfw.minor);
+        let mut major_version = u64::from(self.minfw.major);
+
+        /*
+         * According to the SNP firmware spec, bit 1 of the policy flags is reserved and must
+         * always be set to 1. Rather than passing this responsibility off to callers, set this bit
+         * every time an ioctl is issued to the kernel.
+         */
+        let flags = self.flags.bits | 0b10;
+        let mut flags_64 = u64::from(flags);
+
+        major_version <<= 8;
+        flags_64 <<= 16;
+
+        val |= minor_version;
+        val |= major_version;
+        val |= flags_64;
+        val &= 0x00FFFFFF;
+
+        val
+    }
+}
+
+/// Encapsulates the various data needed to begin the launch process.
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, Deserialize, Serialize)]
+pub struct SnpStart {
+    /// Describes a policy that the AMD Secure Processor will enforce.
+    pub policy: SnpPolicy,
+
+    /// Userspace address of migration agent. Ignored if MA_EN is 0.
+    pub ma_uaddr: u64,
+
+    /// Indicates if this guest is associated with a migration agent. Otherwise 0.
+    pub ma_en: bool,
+
+    /// Indicates that this launch flow is launching an IMI for the purpose of guest-assisted migration.
+    pub imi_en: bool,
+
+    /// Hypervisor provided value to indicate guest OS visible workarounds.The format is hypervisor defined.
+    pub gosvw: [u8; 16],
+}
+
+/// Encapsulates the various data needed to begin the update process.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub struct SnpUpdate {
+    /// Indicates that this page is part of the IMI of the guest.
+    pub imi_page: bool,
+
+    /// Encoded page type.
+    pub page_type: SnpPageType,
+
+    /// VMPL3 permission mask.
+    pub vmpl3_perms: VmplPerms,
+
+    /// VMPL2 permission mask.
+    pub vmpl2_perms: VmplPerms,
+
+    /// VMPL1 permission mask.
+    pub vmpl1_perms: VmplPerms,
+}
+
+bitflags! {
+    #[derive(Default, Deserialize, Serialize)]
+    /// VMPL permission masks.
+    pub struct VmplPerms: u8 {
+        /// Page is readable by the VMPL.
+        const READ =                1;
+
+        /// Page is writeable by the VMPL.
+        const WRITE =               1 << 1;
+
+        /// Page is executable by the VMPL in CPL3.
+        const EXECUTE_USER =        1 << 2;
+
+        /// Page is executable by the VMPL in CPL2, CPL1, and CPL0.
+        const EXECUTE_SUPERVISOR =  1 << 3;
```

Comment:
Weird spacing. Generally when one puts identifiers and values in columns like this, the `=` is one space away from the value.